src/components/shared/thread/ActivityGraphFills.js

```diff
@@ -380,32 +380,49 @@ export class ActivityFillGraphQuerier {
       return null;
     }
 
-    let { offsetToCategoryStart } = categoryUnderMouse;
-    const candidateSamples = this._getCategoriesSamplesAtTime(
-      time,
-      categoryUnderMouse.category
-    );
-
-    for (let i = 0; i < candidateSamples.length; i++) {
-      const { sample, contribution } = candidateSamples[i];
-      if (offsetToCategoryStart <= contribution) {
+    // Get all samples that contribute pixels to the clicked category in this
+    // pixel column of the graph.
+    const { category, categoryLowerEdge, yPercentage } = categoryUnderMouse;
+    const candidateSamples = this._getCategoriesSamplesAtTime(time, category);
+
+    // The candidate samples are sorted by gravity, bottom to top.
+    // Each sample occupies a subrange of the [0, 1] range. The height of each
+    // sample's range is called "contribution" here. The sample ranges are
+    // directly adjacent, there's no space between them.
+    // yPercentage is the mouse position converted to the [0, 1] range. We want
+    // to find the sample whose range contains that yPercentage value.
+    // Since we already filtered the contributing samples by the clicked
+    // category, we start stacking up their contributions onto the lower edge
+    // of that category's fill.
+    let upperEdgeOfPreviousSample = categoryLowerEdge;
+    // Loop invariant: upperEdgeOfPreviousSample <= yPercentage.
+    for (const { sample, contribution } of candidateSamples) {
+      const upperEdgeOfThisSample = upperEdgeOfPreviousSample + contribution;
+      if (yPercentage <= upperEdgeOfThisSample) {
```

Comment:
Small nit: I see you write above  `// Loop invariant: upperEdgeOfPreviousSample <= yPercentage.`, so that would mean that the check here should be `<` (or change the comment above) ? Probably doesn't make a real difference, but this would help understanding the loop.
Also maybe use the same operand order tha nin the comment: `if (upperEdgeOfThisSample > yPercentage)` ? Again the goal is to make it as easy to understand as possible.

Reply:
Hehe, I actually thought about this for longer than I'd like to admit.
The thing that makes this confusing is that we want to return a non-null sample for the values at both edges too, e.g. the values 0.0 and 1.0, not just for values between them (or just including one edge).

Let me start by saying that, as written, the loop invariant is not wrong. If we don't get into the `if (yPercentage <= upperEdgeOfThisSample)` branch, it means that `yPercentage > upperEdgeOfThisSample`. And then after the next statement, `yPercentage > upperEdgeOfPreviousSample`. And if `yPercentage > upperEdgeOfPreviousSample` is true, then `yPercentage >= upperEdgeOfPreviousSample` is also true, so the loop invariant is maintained.
So why didn't I write `yPercentage > upperEdgeOfPreviousSample` for the loop invariant? Because in the *very first* loop iteration, `yPercentage` can actually be `== upperEdgeOfPreviousSample`. This happens if both `yPercentage` and `categoryLowerEdge` are zero, for example.

I'll add more comments that hopefully make this clearer.

Reply:
Oh, and yes, I will invert the directions so that the two are consistent.